zbarimg/zbarimg.c

```diff
@@ -160,13 +160,21 @@ static inline int dump_error(MagickWand *wand)
     return (exit_code);
 }
 
+int file_extension(const char* str, const char* suffix)
+{
+  int l1 = strlen(str);
+  int l2 = strlen(suffix);
+  return strcmp(str + (l1-l2), suffix) == 0;
+}
+
 static int scan_image(const char *filename)
 {
     if (exit_code == 3)
 	return (-1);
 
     int found	       = 0;
     MagickWand *images = NewMagickWand();
+    if (file_extension(filename, ".pdf")) MagickSetResolution(images, 300, 300);
```

Comment:
Maybe replace the `file_extension` check with one based on [`MagickGetImageFormat`](https://imagemagick.org/api/magick-image.php#MagickGetImageFormat).

Reply:
This is a very sensible suggestion, but SetResolution only works _before_ the images are loaded, and GetImageFormat only works _after_ the images are loaded.

I'm open to suggestions on how to do it anyway--I'm pretty much an ImageMagick newbie. Seems like it would be nice to have it work for stdin, for example.

Reply:
(And before you ask, if you just unconditionally set the resolution, it overrides the DPI settings in file formats that actually do set it internally)

Reply:
Hmm, I think you use [`MagickPingImage`](https://imagemagick.org/api/magick-image.php#MagickPingImage) to get the format before doing the full load.

Reply:
Maybe need to remove the image created by ping image using something like [`MagickRemoveImage`](https://imagemagick.org/api/magick-image.php#MagickRemoveImage) before the full load.

Reply:
> Seems like it would be nice to have it work for stdin, for example.

For stdin I think you could use [`MagickReadImageBlob`](https://imagemagick.org/api/magick-image.php#MagickReadImageBlob) and [`MagickPingImageBlob`](https://imagemagick.org/api/magick-image.php#MagickPingImageBlob) for reading/pinging a stdin buffer.

Reply:
I've made the change for a single image. For stdin I don't think this is possible, because you can't seek back to the beginning of stdin--once it's consumed, it's consumed, in general.

Reply:
> For stdin I don't think this is possible, because you can't seek back to the beginning of stdin--once it's consumed, it's consumed, in general.

I think you would read stdin until `EOF` into a dynamically allocated char buffer then pass the buffer to imagemagic using the blob read functions(which shouldn't remove that buffer and thus allow reuse).

Anyways that's mostly separate from this fix at least but may be a nice follow up.

Reply:
Actually, `scan_image` is used for "-" (which I realized as I started said followup). Currently this PR breaks all zbarimg use with stdin. I'll go ahead and figure out a fix. I'll also look into whether there's a missing test--`make check` shows no issues.